driver-core/src/main/com/mongodb/client/model/expressions/SwitchExpression.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import com.mongodb.annotations.Immutable;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.util.Collections.unmodifiableList;
+
+/**
+ * Evaluates a series of case expressions. When it finds an expression which evaluates to true, $switch executes a specified expression
+ * and breaks out of the control flow.
+ *
+ * @since 4.?
+ */
+@Immutable
+public final class SwitchExpression implements Expression {
+    private final List<Branch> branches;
+    @Nullable
+    private final Expression defaultExpr;
+
+    SwitchExpression(final List<Branch> branches, @Nullable final Expression defaultExpr) {
+        this.branches = notNull("branches", branches);
+        this.defaultExpr = defaultExpr;
+    }
+
+    /**
+     * The path to take if no branch case expression evaluates to true.
+     *
+     * @param defaultExpr an expression to evaluate if no branch evaluates to true
+     * @return a new SwitchExpression
+     */
+    public SwitchExpression defaultExpr(final Expression defaultExpr) {
```

Comment:
This is what drove me initially to create concrete classes for each expression.  With this, you can write:

      switchExpr(branch(...), branch(...)).defaultExpr(...)

without it you need to start adding overloads to the builder, e.g.

      Expression switchExpr(Branch ... branch)
      Expression switchExpr(List<Branch> branches)
      Expression switchExpr(List<Branch> branches, Expression defaultExpr)

which seemed like it would create a lot of noise in the builder class.  This pattern is similar to how we define iterable interfaces and options classes, where the required arguments go in the initiating method, and optional arguments are chained.

Reply:
I'd lean towards classes for each expression, for the same reasons you state and the unit tests would end up being more focused on a single expression per test class, which I find cleaner and easier to read.

Reply:
Agreed.